File: services/frontend/src/lib/api/schemas/transaction.ts
import { z } from "zod";

export const TransactionSummarySchema = z.object({
  city_code: z.string(),
  transaction_year: z.number(),
  property_type: z.string(),
  tx_count: z.number(),
  avg_total_price: z.number(),
  median_total_price: z.number(),
  avg_price_sqm: z.number().nullable(),
  avg_area_sqm: z.number().nullable(),
  avg_walk_min: z.number().nullable(),
});

export type TransactionSummary = z.infer<typeof TransactionSummarySchema>;

export const TransactionDetailSchema = z.object({
  city_code: z.string(),
  city_name: z.string(),
  district_name: z.string().nullable(),
  property_type: z.string(),
  total_price: z.number(),
  price_per_sqm: z.number().nullable(),
  area_sqm: z.number().nullable(),
  floor_plan: z.string().nullable(),
  building_year: z.number().nullable(),
  building_structure: z.string().nullable(),
  nearest_station: z.string().nullable(),
  station_walk_min: z.number().nullable(),
  transaction_quarter: z.string(),
});

export type TransactionDetail = z.infer<typeof TransactionDetailSchema>;

File: services/backend/src/domain/transaction.rs
/// 取引価格集計（マテリアライズドビュー由来）。
#[derive(Debug, Clone)]
pub struct TransactionSummary {
    pub city_code: String,
    pub transaction_year: i16,
    pub property_type: String,
    pub tx_count: i32,
    pub avg_total_price: i64,
    pub median_total_price: i64,
    pub avg_price_sqm: Option<i32>,
    pub avg_area_sqm: Option<i32>,
    pub avg_walk_min: Option<i16>,
}

/// 取引価格明細。
#[derive(Debug, Clone)]
pub struct TransactionDetail {
    pub city_code: String,
    pub city_name: String,
    pub district_name: Option<String>,
    pub property_type: String,
    pub total_price: i64,
    pub price_per_sqm: Option<i32>,
    pub area_sqm: Option<i32>,
    pub floor_plan: Option<String>,
    pub building_year: Option<i16>,
    pub building_structure: Option<String>,
    pub nearest_station: Option<String>,
    pub station_walk_min: Option<i16>,
    pub transaction_quarter: String,
}
